File: scala/aoc/src/main/scala/aoc2015/day03_2.scala
package aoc2015

import scala.collection.mutable.HashMap
import scala.collection.mutable.HashSet

case class Point(x: Long, y: Long):
  override def toString() = s"($x, $y)"
  def north = Point(x, y + 1)
  def south = Point(x, y - 1)
  def east = Point(x + 1, y)
  def west = Point(x - 1, y)


def part1(input: String): Long = 
  val dropoffs = HashSet[Point]()
  val source = Point(0, 0)
  dropoffs += source
  var curr = source
  input.foreach:
    c =>
      curr = c match
        case '^' => curr.north
        case 'v' => curr.south
        case '>' => curr.east
        case '<' => curr.west
      dropoffs += curr
  dropoffs.size

def part2(input: String): Long =
  val (santa, robo) = input.zipWithIndex.partition(_._2 % 2 == 0)
  def dropoffs(s: String) =
    val res = HashSet[Point]()
    val source = Point(0, 0)
    res += source
    var curr = source
    s.foreach:
      c =>
        curr = c match
          case '^' => curr.north
          case 'v' => curr.south
          case '>' => curr.east
          case '<' => curr.west
        res += curr
    res
  dropoffs(santa.map(_._1).mkString).union(dropoffs(robo.map(_._1).mkString)).size

@main def main() =
  val input = io.Source.fromFile("input.txt").getLines().mkString("\n").trim()
  println(s"2015 Day 3, Part 1: ${part1(input)}")
  println(s"2015 Day 3, Part 2: ${part2(input)}")

File: haskell/aoc/src/AOC2015/day03_2.hs
import Data.Set (fromList, union)
import Data.Functor ((<&>))
import Data.List (partition)

type Point = (Int, Int)

(<+>) :: Point -> Point -> Point
(<+>) (x1, y1) (x2, y2) = (x1 + x2, y1 + y2)

direction :: Char -> Point
direction '^' = (0, 1)
direction 'v' = (0, -1)
direction '>' = (1, 0)
direction '<' = (-1, 0)
direction _= (0, 0)

part1 :: String -> Int
part1 = length . fromList . scanl (<+>) (0, 0) . fmap direction

zipWithIndex :: [a] -> [(Int, a)]
zipWithIndex = zip [0..]

both :: (a -> b) -> (a, a) -> (b, b)
both f (x, y) = (f x, f y)

f = let (santa, robo) = both (fmap snd) . partition (even . fst) $ zipWithIndex "^>v<"
  in fromList . scanl (<+>) (0, 0) . fmap direction $ santa

part2 :: String -> Int
part2 s = let (santa, robo) = both (fmap snd) . partition (even . fst) $ zipWithIndex s
  in length $ g santa `union` g robo
  where
    g = fromList . scanl (<+>) (0, 0) . fmap direction

main :: IO ()
main = do
  input <- readFile "input.txt"
  putStrLn $ "2015 Day 3, Part 1: " ++ show (part1 input)
  putStrLn $ "2015 Day 3, Part 2: " ++ show (part2 input)
